src/engines/qwant/qwant.go

```diff
@@ -0,0 +1,200 @@
+	}
+	log.Trace().Msgf("%v: UserAgent: %v", seName, options.UserAgent)
+
+	var col *colly.Collector
+	if options.MaxPages == 1 {
+		col = colly.NewCollector(colly.MaxDepth(1), colly.UserAgent(options.UserAgent)) // so there is no thread creation overhead
+	} else {
+		col = colly.NewCollector(colly.MaxDepth(1), colly.UserAgent(options.UserAgent), colly.Async(true))
+	}
+	pagesCol := colly.NewCollector(colly.MaxDepth(1), colly.UserAgent(options.UserAgent), colly.Async(true))
+
+	var retError error
+
+	pagesCol.OnRequest(func(r *colly.Request) {
+		if err := ctx.Err(); err != nil { // dont fully understand this
+			log.Error().Msgf("%v: Pages Collector; Error OnRequest %v", seName, r)
+			r.Abort()
+			retError = err
+			return
+		}
+		r.Ctx.Put("originalURL", r.URL.String())
+	})
+
+	pagesCol.OnError(func(r *colly.Response, err error) {
+		log.Debug().Msgf("%v: Pages Collector; Error OnError:\nURL: %v\nError: %v", seName, r.Ctx.Get("originalURL"), err)
+		log.Trace().Msgf("%v: HTML Response:\n%v", seName, string(r.Body))
+		//retError = err
+	})
+
+	pagesCol.OnResponse(func(r *colly.Response) {
+		urll := r.Ctx.Get("originalURL")
+
+		bucket.SetResultResponse(urll, r, relay, seName)
+	})
+
+	col.OnRequest(func(r *colly.Request) {
+		if err := ctx.Err(); err != nil { // dont fully understand this
+			log.Error().Msgf("%v: SE Collector; Error OnRequest %v", seName, r)
+			r.Abort()
+			retError = err
+			return
+		}
+	})
+
+	col.OnError(func(r *colly.Response, err error) {
+		log.Error().Msgf("%v: SE Collector; Error OnError:\nURL: %v\nError: %v", seName, r.Request.URL.String(), err)
+		log.Trace().Msgf("%v: HTML Response:\n%v", seName, string(r.Body))
+		retError = err
+	})
+
+	col.OnResponse(func(r *colly.Response) {
+		var pageStr string = r.Ctx.Get("page")
+		if pageStr == "" {
+			//the first page
+			return
+		}
+
+		page, _ := strconv.Atoi(pageStr)
+
+		var parsedResponse QwantResponse
+		err := json.Unmarshal(r.Body, &parsedResponse)
+		if err != nil {
+			log.Error().Err(err).Msgf("%v: Failed body unmarshall to json:\n%v", seName, string(r.Body))
+		}
+
+		mainline := parsedResponse.Data.Res.Items.Mainline
+		counter := 0
+		for _, group := range mainline {
+			if group.Type != "web" {
+				continue
+			}
+			for _, result := range group.Items {
+
+				goodURL := utility.ParseURL(result.URL)
+				res := structures.Result{
+					URL:          goodURL,
+					Rank:         -1,
+					SERank:       (page-1)*qResCount + counter,
+					SEPage:       page,
+					SEOnPageRank: counter%defaultResultsPerPage + 1,
+					Title:        result.Title,
+					Description:  result.Description,
+					SearchEngine: seName,
+				}
+				if config.InsertDefaultRank {
+					res.Rank = rank.DefaultRank(res.SERank, res.SEPage, res.SEOnPageRank)
+				}
+
+				bucket.SetResult(&res, relay, options, pagesCol)
+				counter += 1
+			}
+		}
+	})
+
+	col.OnHTML("div[data-testid=\"sectionWeb\"] > div > div", func(e *colly.HTMLElement) {
```

Comment:
When is this used if the API is available?

Reply:
See qwant.md.